agentlightning/tracer/weave.py

```diff
@@ -0,0 +1,335 @@
+        attempt_id: Optional[str] = None,
+    ) -> AsyncContextManager[Any]:
+        """
+        Synchronous implementation of the tracing context.
+
+        Args:
+            name: Optional operation name.
+            store: Optional LightningStore instance.
+            rollout_id: Optional rollout ID.
+            attempt_id: Optional attempt ID.
+
+        Raises:
+            ValueError: If store, rollout_id, and attempt_id are inconsistently provided.
+            RuntimeError: If Weave is not installed or client is uninitialized.
+        """
+        arg_op = name or self.project_name
+        arg_inputs: dict[str, str] | None = {"rollout_id": rollout_id or "", "attempt_id": attempt_id or ""}
+
+        if store is not None and rollout_id is not None and attempt_id is not None:
+            self._rollout_id = rollout_id
+            self._attempt_id = attempt_id
+            self._store = store
+        else:
+            raise ValueError("store, rollout_id, and attempt_id must be either all provided")
+
+        try:
+            import datetime
+
+            import weave
+        except ImportError:
+            raise RuntimeError("Weave is not installed. Install it to use WeaveTracer.")
+
+        weave_client = weave.get_client()  # type: ignore
+        if not weave_client:
+            raise RuntimeError("Weave client is not initialized. Call init_worker() first.")
+
+        # Create a new trace call object in Weave
+        trace_call = weave_client.create_call(op=arg_op, inputs=arg_inputs)  # type: ignore
+        trace_call.started_at = datetime.datetime.now(tz=datetime.timezone.utc)
+
+        try:
+            yield trace_call
+        except Exception as e:
+            # Finish trace and log any exception
+            weave_client.finish_call(trace_call, exception=e)  # type: ignore
+            logger.error(f"Trace failed for rollout_id={rollout_id}, attempt_id={attempt_id}, error={e}")
+        finally:
+            # Finish trace even if no exception
+            weave_client.finish_call(trace_call)  # type: ignore
+            await self._on_finish_handler(trace_call)
+
+    def shutdown(self) -> None:
+        """Stop and clean up the dedicated asyncio loop and thread."""
+        if self._loop:
+            self._loop.call_soon_threadsafe(self._loop.stop)
+            self._loop = None
+        if self._loop_thread:
+            self._loop_thread.join(timeout=5)
+
+    async def _on_finish_handler(self, call: "Call", *args: Any, **kwargs: Any) -> None:  # type: ignore
+        """
+        Handler called when a Weave Call finishes.
+
+        Converts the call (including nested children) into spans and stores them in LightningStore.
+        """
+        spans, self.sequence_id = self.convert_call_to_spans(call, self._rollout_id, self._attempt_id, self.sequence_id)  # type: ignore
+
+        if self._store and self._rollout_id and self._attempt_id:
+            try:
+                for span in spans:
```

Comment:
there is an add_many_spans interface now.

Reply:
Cool, changed